tests/test_config.py

```diff
@@ -0,0 +1,231 @@
+import re
+from pathlib import Path
+
+import pytest
+from pydantic import ValidationError
+
+
+import copier
+from copier.config.factory import make_config
+from copier.config.objects import (
+    DEFAULT_DATA,
+    DEFAULT_EXCLUDE,
+    ConfigData,
+    EnvOps,
+    Flags,
+)
+from copier.config.user_data import (
+    load_config_data,
+    load_json_data,
+    load_toml_data,
+    load_yaml_data,
+    InvalidConfigFileError
+)
+
+
+GOOD_FLAGS = {
+    "pretend": True,
+    "quiet": False,
+    "force": True,
+    "skip": False,
+    "cleanup_on_error": True,
+}
+
+GOOD_ENV_OPS = {
+    "autoescape": True,
+    "block_start_string": "<%",
+    "block_end_string": "%>",
+    "variable_start_string": "<<",
+    "variable_end_string": ">>",
+    "keep_trailing_newline": False,
+    "i_am_not_a_member": None,
+}
+
+
+def test_config_data_is_loaded_from_file():
+    config = load_config_data("tests/demo_data")
+    assert config["_exclude"] == ["exclude1", "exclude2"]
+    assert config["_include"] == ["include1", "include2"]
+    assert config["_skip_if_exists"] == ["skip_if_exists1", "skip_if_exists2"]
+    assert config["_tasks"] == ["touch 1", "touch 2"]
+    assert config["_extra_paths"] == ["tests"]
+
+
+@pytest.mark.parametrize(
+    "template",
+    ["tests/demo_toml", "tests/demo_yaml", "tests/demo_yml", "tests/demo_json"],
+)
+def test_read_data(dst, template):
+    copier.copy(template, dst, force=True)
+
+    gen_file = dst / "user_data.txt"
+    result = gen_file.read_text()
+    expected = Path("tests/user_data.ref.txt").read_text()
+    assert result == expected
+
+
+def test_bad_toml(capsys):
+    with pytest.raises(InvalidConfigFileError):
+        load_toml_data("tests/demo_badtoml")
+
+
+def test_invalid_toml(capsys):
```

Comment:
What is the purposes of `test_invalid_toml` versus `test_bad_toml`?
What is the difference between an _invalid_ and a _bad_ config file in this context?
Why does `test_invalid_toml` call not the loaders for `yaml` and `json` as well?

Reply:
🤔

Reply:
`test_bad_toml` test what happens when `config.toml` is not a valid TOML file.
But It seems it was a mistake adding it, because the badly named `test_invalid_toml` test all different config formats.

TL;DR:
- `test_bad_toml` and the `test/demo_badtoml` foldr should be **removed**.
- `test_invalid_toml`should be **replaced** by the following code

```python
def test_invalid_yaml(capsys):
    assert {} == load_yaml_data("tests/demo_invalid")
    out, _ = capsys.readouterr()
    assert re.search(r"INVALID.*tests/demo_invalid/copier\.yml", out)


def test_invalid_toml(capsys):
    assert {} == load_toml_data("tests/demo_invalid")
    out, _ = capsys.readouterr()
    assert re.search(r"INVALID.*tests/demo_invalid/copier\.toml", out)


def test_invalid_json(capsys):
    assert {} == load_json_data("tests/demo_invalid")
    out, _ = capsys.readouterr()
    assert re.search(r"INVALID.*tests/demo_invalid/copier\.json", out)


def test_invalid_data(capsys):
    assert {} == load_config_data("tests/demo_invalid", _warning=False)
    out, _ = capsys.readouterr()
    assert re.search(r"INVALID", out)
```

Reply:
@jpscaletti Thanks for your input. I am going to make this PR merge ready this weekend.